app/questionnaire/routing/when_rule_evaluator.py

```diff
@@ -0,0 +1,62 @@
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Optional, Union
+
+from app.data_models import AnswerStore, ListStore
+from app.questionnaire import Location, QuestionnaireSchema
+from app.questionnaire.relationship_location import RelationshipLocation
+from app.questionnaire.routing.operator import Operator
+from app.questionnaire.value_source_resolver import (
+    ValueSourceResolver,
+    answer_value_types,
+)
+
+
+@dataclass
+class WhenRuleEvaluator:
+    rule: dict
+    schema: QuestionnaireSchema
+    answer_store: AnswerStore
+    list_store: ListStore
+    metadata: dict
+    location: Union[Location, RelationshipLocation]
+    routing_path_block_ids: Optional[list] = None
+
+    def __post_init__(self) -> None:
+        self.value_source_resolver = ValueSourceResolver(
+            answer_store=self.answer_store,
+            list_store=self.list_store,
+            metadata=self.metadata,
+            schema=self.schema,
+            location=self.location,
+            list_item_id=self.location.list_item_id,
+            routing_path_block_ids=self.routing_path_block_ids,
+            use_default_answer=True,
+        )
+
+    @staticmethod
+    def _is_date_offset(rule: dict) -> bool:
+        return any(x in rule for x in {"days", "months", "years"})
+
+    def _evaluate(
+        self, rule: answer_value_types
+    ) -> Union[bool, datetime, answer_value_types]:
+        if not isinstance(rule, dict) or self._is_date_offset(rule):
+            return rule
+
+        if "source" in rule:
+            return self.value_source_resolver.resolve(rule)
+
+        operator = Operator(next(iter(rule)))
+        operands = rule[operator.name]
+
+        if not isinstance(operands, (list, tuple)):
+            raise TypeError(
+                f"The rule is invalid. Got a non list/tuple for operands - {type(operands)}"
+            )
+
+        operands = (self._evaluate(operand) for operand in operands)
+        return operator.evaluate(operands)
```

Comment:
I find it confusing that `_evaluate` is called when the operand is a value rather than a rule, only to return the value immediately. I expected this to work like the validator code i.e. checking that something is a rule before recursively calling the method.

Reply:
I did that to reduce nesting and have a true recursive nature since the operations are not expensive, plus it returns a generator. Happy to change it to something like this, which only calls evaluate when it needs to resolve an operation and still benefits from a generator.
```py
    def _evaluate(self, rule: answer_value_types):
        operator = Operator(next(iter(rule)))
        operands = rule[operator.name]

        if not isinstance(operands, list):
            raise TypeError(
                f"The rule is invalid. Got a non list type for operands - {type(operands)}"
            )

        resolved_operands = self.get_resolved_operands(operands)
        return operator.evaluate(resolved_operands)

    def get_resolved_operands(self, operands: list):
        for operand in operands:
            if isinstance(operand, dict) and not self._is_date_offset(operand):
                if "source" in operand:
                    yield self.value_source_resolver.resolve(operand)
                else:
                    yield self._evaluate(operand)
            else:
                yield operand
```

Reply:
This is much more readable 👍 

> I didn't explicitly check for operator types, as that isn't strictly necessary.

My concern with this is:
- logic about a specific operator has leaked into this code
- as more operators get added this could overloaded with checks

Reply:
Updated to the above snippet while also explicitly checking Operator types for the reasons you highlighted above.

Reply:
I can't see the change to explicitly check operator types.